File: countSortedVowelStrings.go
// problem:https://leetcode.com/problems/count-sorted-vowel-strings/
// Runtime: 0 ms, faster than 100.00% of Go online submissions for Count Sorted Vowel Strings.
// Memory Usage: 1.9 MB, less than 100.00% of Go online submissions for Count Sorted Vowel Strings.

package algorithms

func combinations(n uint64, r uint64) int {
	divisor, dividend := uint64(1), uint64(1)
	for i := uint64(n - r + 1); i <= n; i++ {
		dividend *= i
	}
	for i := uint64(1); i <= r; i++ {
		divisor *= i
	}
	return int(dividend / divisor)
}

func countVowelStrings(n int) int {
	return combinations(uint64(n+4), uint64(4))
}

File: countSortedVowelStrings.cpp
// problem: https://leetcode.com/problems/count-sorted-vowel-strings/
// Runtime: 0 ms, faster than 100.00% of C++ online submissions for Count Sorted Vowel Strings.
// Memory Usage: 5.8 MB, less than 96.91% of C++ online submissions for Count Sorted Vowel Strings.

typedef unsigned long long ll;

class Solution
{
public:
    int combination(ll n, ll r)
    {
        ll divisor = 1, dividend = 1;
        for (ll i = n - r + 1; i <= n; ++i)
        {
            dividend *= i;
        }
        for (ll i = 1; i <= r; ++i)
        {
            divisor *= i;
        }
        return (int)(dividend / divisor);
    }

    int countVowelStrings(int n)
    {
        return combination((ll)(n + 4), (ll)(4));
    }
};